packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -164,49 +193,87 @@ export class ConnectionStateHandler {
         this._pendingClientId = details.clientId;
 
-        // Report telemetry after we set client id, but before transitioning to Connected state below!
-        this.handler.logConnectionStateChangeTelemetry(ConnectionState.Connecting, oldState);
+        // We will want to catch up to known ops as of now before transitioning to Connected state
+        this.catchUpMonitor = this.handler.createCatchUpMonitor();
 
-        const quorumClients = this.handler.quorumClients();
-        // Check if we already processed our own join op through delta storage!
-        // we are fetching ops from storage in parallel to connecting to ordering service
-        // Given async processes, it's possible that we have already processed our own join message before
+        // This pending clientId could be in the quorum already (i.e. join op already processed).
+        // We are fetching ops from storage in parallel to connecting to Relay Service,
+        // and given async processes, it's possible that we have already processed our own join message before
         // connection was fully established.
-        // Note that we might be still initializing quorum - connection is established proactively on load!
-        if ((quorumClients !== undefined && quorumClients.getMember(details.clientId) !== undefined)
-            || connectionMode === "read"
-        ) {
-            assert(!this.prevClientLeftTimer.hasTimer, 0x2a6 /* "there should be no timer for 'read' connections" */);
-            this.setConnectionState(ConnectionState.Connected);
-        } else if (connectionMode === "write") {
+        const pendingClientAlreadyInQuorum = quorumClients?.getMember(this._pendingClientId) !== undefined;
+
+        // IMPORTANT: Report telemetry after we set _pendingClientId, but before transitioning to Connected state
+        this.handler.logConnectionStateChangeTelemetry(ConnectionState.Connecting, oldState);
+
+        // Prepare to transition to Connected state, which may happen elsewhere once all preconditions are met
+        if (writeConnection && !pendingClientAlreadyInQuorum) {
+            // Previous client left, and we are waiting for our own join op. When it is processed we'll join the quorum
+            // and attempt to transition to Connected state via receivedAddMemberEvent.
             this.startJoinOpTimer();
+        } else if (this.prevClientLeftTimer.hasTimer) {
+            // Nothing to do now - when the previous client is removed from the quorum
+            // we will attempt to transition to Connected state via receivedRemoveMemberEvent
+            assert(writeConnection, 0x2a6 /* "There should be no timer for 'read' connections" */);
+        } else {
+            // We're not waiting for Leave or Join op, but we do need to wait until we are caught up (to now-known ops)
+            // before transitioning to Connected state.
+            this.catchUpMonitor.on("caughtUp", this.transitionToConnectedState);
```

Comment:
Is it worth turning this magic string into an exported variable?

Reply:
No need, It's protected by the type system 😎